modules/grandpa/src/lib.rs

```diff
@@ -319,6 +319,16 @@ pub mod pallet {
 	pub type BestFinalized<T: Config<I>, I: 'static = ()> =
 		StorageValue<_, BridgedBlockId<T, I>, OptionQuery>;
 
+	/// Justification of the best finalized header.
+	///
+	/// NOTE: This storage entry is only used by relayers to check for equivocations.
```

Comment:
IMO events are exactly for that - for stuff that should never be read by runtime code && only accessible outside. I guess you're adding it here as a separate storage item because maybe it is easier to write the relayer code - you don't need to deal with pallet inidices? Or is there some other reason?

I suggest to move it to events. I could imagine for example a batch call with `All(submit_finality_proof(BadProof), submit_finality_proof(GoodProof))` and if we'll only be looking at the state at the end of block, we may miss the `BadProof`. The same with the scheduled calls or XCM `Transact` calls - they can be executed from `on_idle`/`on_initialize` and we may miss something. System events will have a record for every justification of a block && we can easily get all submitted justifications by iterating events.

Reply:
> I could imagine for example a batch call with `All(submit_finality_proof(BadProof), submit_finality_proof(GoodProof))` and if we'll only be looking at the state at the end of block, we may miss the `BadProof`.

Oh, you're right. Nice catch !



> I guess you're adding it here as a separate storage item because maybe it is easier to write the relayer code - you don't need to deal with pallet inidices? Or is there some other reason?

High level, yes, I think it would be easier to write the relayer code. Because we could use [`subscribeStorage`](https://polkadot.js.org/docs/substrate/rpc/#subscribestoragekeys-vecstoragekey-storagechangeset) and get a notification each time `BestJustification` changes. With events from what I understand we can't have a subscription for a certain event. We just have to iterate over all the events each block and look for the ones that we need. I'll do some experiments. But that's what I understood from the documentation.

Reply:
> High level, yes, I think it would be easier to write the relayer code. Because we could use [subscribeStorage](https://polkadot.js.org/docs/substrate/rpc/#subscribestoragekeys-vecstoragekey-storagechangeset) and get a notification each time BestJustification changes. With events from what I understand we can't have a subscription for a certain event. We just have to iterate over all the events each block and look for the ones that we need. I'll do some experiments. But that's what I understood from the documentation.

True, but I'm not sure how this RPC subscription works with multiple changes ar the single block. E.g. the same `All(submit_finality_proof(BadProof), submit_finality_proof(GoodProof))` - it'll emit the single event with `GoodProof` or two with `BadProof` and `GoodProof`. I think the latter, because it is implemented at the client level so it doesn't know what happens inside the block. I may be wrong here, though.

Also - I wouldn't rely on subscriptions - it should be combined with simple storage reads. Subscriptions have some caveats, that maybe are ok for regular usages, but for something that you're working on (i.e. component that can't tolerate any missing justifications). Like IIRC e.g. finality subscription may skip ranges of blocks if it finalizes many blocks at once. Also I'm not sure if there's a limit on number of active subscriptions on a node? Also - regular network outages, which can lead to missing blocks => justifications.

So imo your component may use subscriptions, but it should also track latest block where you have read justifications. And if something goes wrong (network or process crash), it should iterate all missed blocks on restart and check justifications from there. Maybe it'd be easier to avoid subscriptions at all and use regular polling, but it is up to you

Reply:
Done. PTAL !

Reply:
Sorry for the delay - I promise to review tomorrow